app/src/main/java/edu/wpi/first/shuffleboard/app/plugin/PluginLoader.java

```diff
@@ -0,0 +1,124 @@
+package edu.wpi.first.shuffleboard.app.plugin;
+
+import edu.wpi.first.shuffleboard.api.data.DataTypes;
+import edu.wpi.first.shuffleboard.api.plugin.Plugin;
+import edu.wpi.first.shuffleboard.api.sources.SourceTypes;
+import edu.wpi.first.shuffleboard.api.sources.recording.serialization.Serializers;
+import edu.wpi.first.shuffleboard.api.widget.Widgets;
+
+import java.io.File;
+import java.lang.reflect.Modifier;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.jar.JarFile;
+import java.util.stream.Stream;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+
+public class PluginLoader {
+
+  private static final PluginLoader defaultLoader = new PluginLoader();
+
+  private final ObservableList<Plugin> knownPlugins = FXCollections.observableArrayList();
+
+  public static PluginLoader getDefault() {
+    return defaultLoader;
+  }
+
+  /**
+   * Loads a plugin jar and loads all plugin classes within.
+   *
+   * @param jarFile the jar file to load plugins from
+   */
+  public void loadPluginJar(JarFile jarFile) throws MalformedURLException {
+    URL url = new File(jarFile.getName()).toURI().toURL();
+    URLClassLoader classLoader = new URLClassLoader(new URL[]{url}, ClassLoader.getSystemClassLoader());
+
+    jarFile.stream()
+        .filter(e -> e.getName().endsWith(".class"))
+        .map(e -> e.getName().replace('/', '.'))
+        .map(n -> n.substring(0, n.length() - 6)) // ".class".length() == 6
+        .flatMap(className -> tryLoadClass(className, classLoader))
+        .forEach(this::loadPluginClass);
+  }
+
+  private Stream<Class<?>> tryLoadClass(String name, ClassLoader classLoader) {
+    try {
+      return Stream.of(Class.forName(name, false, classLoader));
+    } catch (ClassNotFoundException e) {
+      e.printStackTrace();
+      return Stream.empty();
+    }
+  }
+
+  /**
+   * Attempts to load a plugin class. This class may or not be a plugin; only a plugin class will be loaded.
+   *
+   * @param clazz the class to attempt to load
+   *
+   * @return true if the class is a plugin class and was successfully loaded; false otherwise
+   */
+  public boolean loadPluginClass(Class<?> clazz) {
+    if (Plugin.class.isAssignableFrom(clazz) && !Modifier.isAbstract(clazz.getModifiers())) {
+      try {
+        if (Modifier.isPublic(clazz.getConstructor().getModifiers())) {
+          // Unload existing plugin, if it exists
+          knownPlugins.stream()
+              .filter(p -> p.getClass().getName().equals(clazz.getName()))
+              .forEach(this::unload);
+          Plugin plugin = (Plugin) clazz.newInstance();
+          load(plugin);
+          return true;
+        }
+      } catch (ReflectiveOperationException e) {
+        // TODO log it?
+        //e.printStackTrace();
+        return false;
+      }
+    }
+    return false;
```

Comment:
That'll be logged a lot; one statement per non-plugin class. For example, the `base` plugin jar would log ~20 times here